File: Queue/myqueue.py

import sys
sys.path.append("../Array")
sys.path.append("../LinkedList")




import myarray
import linkedlist


class Queue :
    def __init__(self, size=20) :
        self.m_arr = myarray.Arr(size)
        for i in range(size) :
            self.m_arr.Append(0)

        self.m_head, self.m_tail = 0, 0
        self.m_num = size

    def Enqueue(self, value) :
        if self.m_tail == self.m_num and self.m_head == 0 :
            return False

        if self.m_tail == self.m_num and self.m_head != 0 :
            for i in range(self.m_head, self.m_tail) :
                self.m_arr[i-self.m_head] = self.m_arr[i]

            self.m_tail = self.m_tail - self.m_head
            self.m_head = 0


        self.m_arr[self.m_tail] = value
        self.m_tail = self.m_tail + 1
        return True

    def Dequeue(self) :
        if self.m_head == self.m_tail :
            return False

        #res = self.m_arr[self.m_head]
        self.m_head = self.m_head + 1
        return True


    
    def Print(self) :
        for i in range(self.m_head, self.m_tail) :
            print(self.m_arr[i])



class LinkedQueue :
    def __init__(self) :
        self.m_head = None
        self.m_tail = None

    def Enqueue(self, value) :
        new_node = linkedlist.LinkedNode(value)
        if self.m_tail :
            self.m_tail.next = new_node
        else :
            self.m_head = new_node

        self.m_tail = new_node
        return True

    def Dequeue(self) :
        if not self.m_head :
            return False

        self.m_head = self.m_head.next
        if not self.m_head :
            self.m_tail = None

        return True

    def Print(self) :
        val = []
        p = self.m_head
        while p :
            val.append(p.data)
            p = p.next

        print(val)



class RingQueue :
    def __init__(self, size = 20) :
        self.m_arr = myarray.Arr(size)
        for i in range(size) :
            self.m_arr.Append(None)

        self.m_num = size
        self.m_head, self.m_tail = 0, 0

    def Enqueue(self, value) :
        if (self.m_tail + 1) % self.m_num == self.m_head :
            return False

        self.m_arr[self.m_tail] = value
        self.m_tail = (self.m_tail + 1) % self.m_num
        return True


    def Dequeue(self) :
        if self.m_head == self.m_tail :
            return False

        # res = self.m_arr[self.m_head]
        self.m_head = (self.m_head + 1) % self.m_num
        return True

    
    def Print(self) :
        if self.m_head == self.m_tail :
            return
        
        val = []
        pos = self.m_head
        while pos != self.m_tail :
            val.append(self.m_arr[pos])
            pos = (pos + 1) % self.m_num

        print(val)



if __name__ == "__main__" :

    q = RingQueue(5)
    q.Enqueue(1)
    q.Enqueue(2)
    q.Enqueue(3)
    q.Enqueue(4)
    q.Enqueue(5)
    
    q.Dequeue()
    q.Dequeue()
    q.Dequeue()

    q.Enqueue(6)
    q.Enqueue(7)


    q.Print()

File: Queue/go/myqueue/myqueue.go
package myqueue

import (
	"array"
	"fmt"
	"link"
)

type Queue struct {
	arr  *array.Arr
	head int
	tail int
	num  int
}

func NewQueue(size int) *Queue {
	arr := array.NewArr(size)
	for i := 0; i < size; i++ {
		arr.Append(0)
	}

	q := &Queue{
		arr:  arr,
		head: 0,
		tail: 0,
		num:  size,
	}

	return q
}

func (q *Queue) Enqueue(value int) bool {
	if q.head == 0 && q.tail == q.num {
		return false
	}

	if q.head != 0 && q.tail == q.num {
		for i := q.head; i < q.tail; i++ {
			res, _ := q.arr.GetArr(i)
			q.arr.SetArr(i-q.head, res)
		}

		q.tail = q.tail - q.head
		q.head = 0
	}

	q.arr.SetArr(q.tail, value)
	q.tail++

	return true
}

func (q *Queue) Dequeue() bool {
	if q.head == q.tail {
		return false
	}

	q.head++
	return true
}

func (q *Queue) Print() {
	for i := q.head; i < q.tail; i++ {
		fmt.Println(q.arr.GetArr(i))
	}
}

type LinkedQueue struct {
	head *link.Linkednode
	tail *link.Linkednode
}

func NewLinkedQueue() *LinkedQueue {
	lq := &LinkedQueue{
		head: nil,
		tail: nil,
	}

	return lq
}

func (lq *LinkedQueue) Enqueue(value int) bool {
	new_node := &link.Linkednode{Next: nil, Data: value}
	if lq.tail != nil {
		lq.tail.Next = new_node
	}

	if lq.head == nil {
		lq.head = new_node
	}

	lq.tail = new_node

	return true
}

func (lq *LinkedQueue) Dequeue() bool {
	if lq.head == nil {
		return false
	}

	lq.head = lq.head.Next
	if lq.head == nil {
		lq.tail = nil
	}

	return true
}

func (lq *LinkedQueue) Print() {
	p := lq.head
	for {
		if p == nil {
			break
		}

		fmt.Println(p.Data)
		p = p.Next
	}
}

type RingQueue struct {
	arr  *array.Arr
	head int
	tail int
	num  int
}

func NewRingQueue(size int) *RingQueue {
	arr := array.NewArr(size)
	for i := 0; i < size; i++ {
		arr.Append(0)
	}

	rq := &RingQueue{
		arr:  arr,
		head: 0,
		tail: 0,
		num:  size,
	}

	return rq
}

func (rq *RingQueue) Enqueue(value int) bool {
	if (rq.tail+1)%rq.num == rq.head {
		return false
	}

	rq.arr.SetArr(rq.tail, value)
	rq.tail = (rq.tail + 1) % rq.num
	return true
}

func (rq *RingQueue) Dequeue() bool {
	if rq.head == rq.tail {
		return false
	}

	rq.head = (rq.head + 1) % rq.num
	return true
}

func (rq *RingQueue) Print() {
	pos := rq.head
	for {
		if pos == rq.tail {
			break
		}

		fmt.Println(rq.arr.GetArr(pos))
		pos = (pos + 1) % rq.num
	}
}
